packages/cli/src/commands/base-command.ts

```diff
@@ -71,8 +72,24 @@ export abstract class BaseCommand extends Command {
 
 	protected async loadModules() {
 		for (const moduleName of this.modulesConfig.modules) {
-			await import(`../modules/${moduleName}/${moduleName}.module`);
-			this.logger.debug(`Loaded module "${moduleName}"`);
+			let preInitModule: ModulePreInit | undefined;
+			try {
+				preInitModule = (await import(
+					`../modules/${moduleName}/${moduleName}.pre-init`
+				)) as ModulePreInit;
+			} catch (e) {}
+
+			if (
+				!preInitModule ||
+				preInitModule.shouldLoadModule?.({
+					database: this.globalConfig.database,
+					instance: this.instanceSettings,
+				})
+			) {
+				await import(`../modules/${moduleName}/${moduleName}.module`);
```

Comment:
A bit sick today so maybe a stupid question, but how does this work? I'd expect we need to do more than merely importing, namely instantiate the module class so that the background process can start running? Or `InsightsModule.constructor` gets called in some way I'm not aware of?

Reply:
The module gets instantiated here: https://github.com/n8n-io/n8n/blob/master/packages/cli/src/decorators/module.ts#L22 when we register the hook handlers (the DI gets called and instantiate the module)
But it could make sense to initialize it before :thinking:

Reply:
Discussed over a call, it turns out compaction is only starting when an execution finishes instead of on n8n startup. We'll wait for this fix.

Reply:
So the way it works now is that the import adds the module to both the DI system and the ModuleRegistry service (thanks to decorators). Kind of how we do it with Controllers 
The module instance should not thus be directly instantiated (because the DI system should be responsible for managing the instance).
I added an `initializeModules` function on the ModuleRegistry to call the initialize method on all the registered modules that implement it. I call this `initializeModules` function on loadModules command after all modules have been loaded